File: Habitica/src/main/java/com/magicmicky/habitrpgwrapper/lib/models/tasks/Days.java
package com.magicmicky.habitrpgwrapper.lib.models.tasks;

import android.os.Parcel;
import android.os.Parcelable;

import com.habitrpg.android.habitica.HabitDatabase;
import com.raizlabs.android.dbflow.annotation.Column;
import com.raizlabs.android.dbflow.annotation.ModelContainer;
import com.raizlabs.android.dbflow.annotation.NotNull;
import com.raizlabs.android.dbflow.annotation.PrimaryKey;
import com.raizlabs.android.dbflow.annotation.Table;
import com.raizlabs.android.dbflow.structure.BaseModel;

/**
 * Created by viirus on 06/07/15.
 */
@ModelContainer
@Table(databaseName = HabitDatabase.NAME)
public class Days extends BaseModel implements Parcelable {

    @Column
    @PrimaryKey
    @NotNull
    String task_id;

    @Column
    private boolean m, t, w, th, f, s, su;

    public Days() {
        this.m = false;
        this.t = false;
        this.w = false;
        this.th = false;
        this.f = false;
        this.s = true;
        this.su = true;
    }

    public boolean getT() {
        return t;
    }

    public void setT(boolean t) {
        this.t = t;
    }

    public boolean getW() {
        return w;
    }

    public void setW(boolean w) {
        this.w = w;
    }

    public boolean getTh() {
        return th;
    }

    public void setTh(boolean th) {
        this.th = th;
    }

    public boolean getF() {
        return f;
    }

    public void setF(boolean f) {
        this.f = f;
    }

    public boolean getS() {
        return s;
    }

    public void setS(boolean s) {
        this.s = s;
    }

    public boolean getSu() {
        return su;
    }

    public void setSu(boolean su) {
        this.su = su;
    }

    public boolean getM() {
        return m;
    }

    public void setM(boolean m) {
        this.m = m;
    }

    public boolean getForDay(int day) {
        switch (day) {
            case 2:
                return this.getM();
            case 3:
                return this.getT();
            case 4:
                return this.getW();
            case 5:
                return this.getTh();
            case 6:
                return this.getF();
            case 7:
                return this.getS();
            case 1:
                return this.getSu();
        }
        return false;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.task_id);
        dest.writeByte(this.m ? (byte) 1 : (byte) 0);
        dest.writeByte(this.t ? (byte) 1 : (byte) 0);
        dest.writeByte(this.w ? (byte) 1 : (byte) 0);
        dest.writeByte(this.th ? (byte) 1 : (byte) 0);
        dest.writeByte(this.f ? (byte) 1 : (byte) 0);
        dest.writeByte(this.s ? (byte) 1 : (byte) 0);
        dest.writeByte(this.su ? (byte) 1 : (byte) 0);
    }

    protected Days(Parcel in) {
        this.task_id = in.readString();
        this.m = in.readByte() != 0;
        this.t = in.readByte() != 0;
        this.w = in.readByte() != 0;
        this.th = in.readByte() != 0;
        this.f = in.readByte() != 0;
        this.s = in.readByte() != 0;
        this.su = in.readByte() != 0;
    }

    public static final Parcelable.Creator<Days> CREATOR = new Parcelable.Creator<Days>() {
        @Override
        public Days createFromParcel(Parcel source) {
            return new Days(source);
        }

        @Override
        public Days[] newArray(int size) {
            return new Days[size];
        }
    };
}

File: shared/src/androidMain/kotlin/com/habitrpg/shared/habitica/models/tasks/Days.kt
package com.habitrpg.shared.habitica.models.tasks

import android.os.Parcel
import android.os.Parcelable

import io.realm.RealmObject
import io.realm.annotations.PrimaryKey

actual open class Days : RealmObject, Parcelable {

    @PrimaryKey
    actual var taskId: String? = null
    actual var m: Boolean = false
    actual var t: Boolean = false
    actual var w: Boolean = false
    actual var th: Boolean = false
    actual var f: Boolean = false
    actual var s: Boolean = false
    actual var su: Boolean = false

    actual fun getForDay(day: Int): Boolean {
        when (day) {
            2 -> return this.m
            3 -> return this.t
            4 -> return this.w
            5 -> return this.th
            6 -> return this.f
            7 -> return this.s
            1 -> return this.su
        }
        return false
    }

    actual override fun describeContents(): Int {
        return 0
    }

    actual override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeString(this.taskId)
        dest.writeByte(if (this.m) 1.toByte() else 0.toByte())
        dest.writeByte(if (this.t) 1.toByte() else 0.toByte())
        dest.writeByte(if (this.w) 1.toByte() else 0.toByte())
        dest.writeByte(if (this.th) 1.toByte() else 0.toByte())
        dest.writeByte(if (this.f) 1.toByte() else 0.toByte())
        dest.writeByte(if (this.s) 1.toByte() else 0.toByte())
        dest.writeByte(if (this.su) 1.toByte() else 0.toByte())
    }

    protected actual constructor(`in`: Parcel) {
        this.taskId = `in`.readString()
        this.m = `in`.readByte().toInt() != 0
        this.t = `in`.readByte().toInt() != 0
        this.w = `in`.readByte().toInt() != 0
        this.th = `in`.readByte().toInt() != 0
        this.f = `in`.readByte().toInt() != 0
        this.s = `in`.readByte().toInt() != 0
        this.su = `in`.readByte().toInt() != 0
    }

    constructor() {
        m = true
        t = true
        w = true
        th = true
        f = true
        s = true
        su = true
    }

    companion object {

        @JvmField
        val CREATOR: Parcelable.Creator<Days> = object : Parcelable.Creator<Days> {
            override fun createFromParcel(source: Parcel): Days {
                return Days(source)
            }

            override fun newArray(size: Int): Array<Days?> {
                return arrayOfNulls(size)
            }
        }
    }
}
